numpy/core/tests/test_multiarray.py

```diff
@@ -818,6 +818,30 @@ def __array__(self, dtype=None):
 
         assert_raises(ValueError, np.array, x())
 
+    def test_bad_array_attribute_and_length(self):
+        # See gh-17785, pre 1.19.x this created an array, this was changed
+        # but then lead to stack overflows or hanging.
+
+        class MyArray(np.ndarray):
+            def __new__(cls, input_array):
+                obj = np.empty(len(input_array), dtype=object)
+                obj[:] = input_array
+                obj = obj.view(cls)
+                return obj
+
+        class MyNastyClass:
+            def __len__(self): return 32
+            def __getitem__(self, item): return 1
+            def __array__(self): return MyArray([self])
+
+        try:
+            # This should raise a RecursionError, but sometimes does not
+            # (for reasons unclear to me @seberg) on all Python versions/
+            # However, it should definitely not crash.
+            np.array(MyNastyClass())
+        except:
```

Comment:
```suggestion
        except Exception:
```
so as not to catch `KeyboardInterrupt`